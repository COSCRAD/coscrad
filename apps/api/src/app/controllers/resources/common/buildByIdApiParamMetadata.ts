export default () => ({
    name: 'id',
    required: true,
    example: '2',
});
