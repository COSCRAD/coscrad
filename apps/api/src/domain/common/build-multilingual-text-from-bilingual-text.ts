import { InternalError } from '../../lib/errors/InternalError';
import { isNullOrUndefined } from '../utilities/validation/is-null-or-undefined';
import { buildMultilingualTextWithSingleItem } from './build-multilingual-text-with-single-item';
import {
    MultilingualText,
    MultilingualTextItem,
    MultilingualTextItemRole,
} from './entities/multilingual-text';

export class InvalidBilingualTextException extends InternalError {
    constructor() {
        super(
            `Bilingual text must have non-empty text for either the primary or secondary language`
        );
    }
}

export type TextAndLanguageCode = Pick<MultilingualTextItem, 'text' | 'languageCode'>;

export const buildMultilingualTextFromBilingualText = (
    primary: TextAndLanguageCode,
    secondary: TextAndLanguageCode
) => {
    if (isNullOrUndefined(primary) || isNullOrUndefined(secondary))
        throw new InvalidBilingualTextException();

    if (isNullOrUndefined(primary.text)) {
        if (isNullOrUndefined(secondary.text)) throw new InvalidBilingualTextException();

        return buildMultilingualTextWithSingleItem(secondary.text, secondary.languageCode);
    }

    const multilingualTextForPrimary = buildMultilingualTextWithSingleItem(
        primary.text,
        primary.languageCode
    );

    return isNullOrUndefined(secondary.text)
        ? multilingualTextForPrimary
        : (multilingualTextForPrimary.translate(
              new MultilingualTextItem({
                  ...secondary,
                  role: MultilingualTextItemRole.freeTranslation,
              })
          ) as MultilingualText);
};
