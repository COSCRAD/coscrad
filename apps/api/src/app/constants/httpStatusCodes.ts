import { ValueType } from '../../lib/types/valueType';

const httpStatusCodes = {
    ok: 200,
    badRequest: 400,
    unauthorized: 401,
    forbidden: 403,
    notFound: 404,
    internalError: 500,
};

export default httpStatusCodes;

export type HttpStatusCode = ValueType<typeof httpStatusCodes>;
