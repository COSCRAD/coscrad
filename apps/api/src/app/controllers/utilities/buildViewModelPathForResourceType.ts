import { ResourceType, resourceTypes } from '../../../domain/types/resourceTypes';

// For irregular plural forms or forms that lead to awkward grammar
const exceptions: Partial<Record<ResourceType, string>> = {
    [resourceTypes.transcribedAudio]: 'transcribedAudioItems',
};

export default (resourceType: ResourceType): string =>
    exceptions[resourceType] || `${resourceType}s`;
