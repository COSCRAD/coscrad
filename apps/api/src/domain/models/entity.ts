import { PartialDTO } from 'apps/api/src/types/partial-dto';
import { EntityType } from '../types/entityType';
import { isNullOrUndefined } from '../utilities/validation/is-null-or-undefined';
import { EntityId } from './types/entity-id';
import { EntityCompositeIdentifier } from './types/entityCompositeIdentifier';

// TODO Extract to a separate location
const isValidStringWithLength = (input: unknown): input is string =>
  typeof input === 'string' && input.length > 0;

/**
 *
 * TODO move this near the corresponding type. Note that this extra layer is
 * meant to safe guard us in case we want stricter validation on an `entity id`.
 * Should we change the rule, we can update this in once place, instead of replacing
 * `isValidStringWithLength` everywhere.
 */
const isValidEntityId = (input: unknown): input is EntityId =>
  isValidStringWithLength(input);

export abstract class Entity {
  readonly id: EntityId;

  abstract readonly type: EntityType;

  readonly published: boolean;

  constructor(dto: unknown) {
    // TODO Validate DTO
    if (!this.#validateDTO(dto))
      throw new Error(`Invalid dto for an entity: ${JSON.stringify(dto)}`);

    this.id = dto.id;

    this.published = typeof dto.published === 'boolean' ? dto.published : false;
  }

  getCompositeIdentifier = (): EntityCompositeIdentifier => ({
    type: this.type,
    id: this.id,
  });

  /**
   * TODO consider moving this to a `Serializable` mixin. There may be cases
   * where we want this behaviour on a non domain-model class without the
   * inheritance baggage.
   *  */
  toDTO<TEntity extends Entity>(this: TEntity): PartialDTO<TEntity> {
    return JSON.parse(JSON.stringify(this));
  }

  /**
   * TODO develop a general pattern \ directory structure for validation \ validators.
   * One idea: Return a symbol Valid | Errors[] from a validator instead of a
   * boolean flag.
   *
   * We may break the validators out into separate files. Then we will validate
   * a DTO before passing it to the constructor in our `instance factory` methods
   * in the mapping layer for the repositories.
   *  */
  #validateDTO = (dto: unknown): dto is PartialDTO<Entity> => {
    const { id } = dto as PartialDTO<Entity>;

    // A new entity will have id auto assigned by db
    if (!isNullOrUndefined(id) && !isValidEntityId(id)) return false;

    return true;
  };
}
