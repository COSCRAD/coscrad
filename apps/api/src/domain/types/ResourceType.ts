import { UuidDocument } from '../../lib/id-generation/types/UuidDocument';
import { IBibliographicReference } from '../models/bibliographic-reference/interfaces/bibliographic-reference.interface';
import { Book } from '../models/book/entities/book.entity';
import { Category } from '../models/categories/entities/category.entity';
import { EdgeConnection } from '../models/context/edge-connection.entity';
import { MediaItem } from '../models/media-item/entities/media-item.entity';
import { Photograph } from '../models/photograph/entities/photograph.entity';
import { Song } from '../models/song/song.entity';
import { ISpatialFeature } from '../models/spatial-feature/ISpatialFeature';
import { Tag } from '../models/tag/tag.entity';
import { Term } from '../models/term/entities/term.entity';
import { TranscribedAudio } from '../models/transcribed-audio/entities/transcribed-audio.entity';
import { CoscradUserGroup } from '../models/user-management/group/entities/coscrad-user-group.entity';
import { CoscradUser } from '../models/user-management/user/entities/user/coscrad-user.entity';
import { VocabularyList } from '../models/vocabulary-list/entities/vocabulary-list.entity';
import { AggregateId } from './AggregateId';

export enum ResourceType {
    term = 'term',
    vocabularyList = 'vocabularyList',
    transcribedAudio = 'transcribedAudio',
    book = 'book',
    photograph = 'photograph',
    spatialFeature = 'spatialFeature',
    bibliographicReference = 'bibliographicReference',
    song = 'song',
    mediaItem = 'mediaItem',
}

export const isResourceType = (input: unknown): input is ResourceType =>
    Object.values(ResourceType).includes(input as ResourceType);

// We should use this for type inference a few places.
export type ResourceTypeToResourceModel = {
    term: Term;
    vocabularyList: VocabularyList;
    transcribedAudio: TranscribedAudio;
    book: Book;
    photograph: Photograph;
    spatialFeature: ISpatialFeature;
    bibliographicReference: IBibliographicReference;
    song: Song;
    mediaItem: MediaItem;
};

/**
 * This represents the state of all domain models, excluding their `Connections`
 */
export type InMemorySnapshotOfResources = {
    [K in ResourceType]?: ResourceTypeToResourceModel[K][];
};

export type InMemorySnapshot = {
    resources: InMemorySnapshotOfResources;
    connections: EdgeConnection[];
    tags: Tag[];
    /**
     * We do not intend to leak the abstraction of how the categories are
     * represented in the database here. Defer this to (the db specific) document
     * mapping layer.
     */
    categoryTree: Category[];
    users: CoscradUser[];
    userGroups: CoscradUserGroup[];
    uuids: UuidDocument<AggregateId>[];
};
