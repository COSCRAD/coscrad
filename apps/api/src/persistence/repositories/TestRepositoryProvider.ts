import { Entity } from '../../domain/models/entity';
import { EntityType, InMemorySnapshot } from '../../domain/types/entityType';
import { DatabaseProvider } from '../database/database.provider';
import { getArangoCollectionIDFromEntityType } from '../database/get-arango-collection-ids';
import { RepositoryProvider } from './repository.provider';

export default class TestRepositoryProvider extends RepositoryProvider {
  constructor(databaseProvider: DatabaseProvider) {
    super(databaseProvider);
  }

  // TODO We should correlate entity type with TEntity here
  public async addEntitiesOfSingleType<TEntity extends Entity>(
    entityType: EntityType,
    entities: TEntity[]
  ): Promise<void> {
    await this.forEntity<TEntity>(entityType).createMany(entities);
  }

  // TODO fix types
  public async addEntitiesOfManyTypes(
    snapshot: InMemorySnapshot
  ): Promise<void> {
    const writePromises = Object.entries(snapshot).map(
      ([entityType, entityInstances]) =>
        this.addEntitiesOfSingleType(
          entityType as EntityType,
          entityInstances as any
        )
    );
  }

  public async deleteAllEntitiesOfGivenType(
    entityType: EntityType
  ): Promise<void> {
    await (
      await this.databaseProvider.getDBInstance()
    ).deleteAll(getArangoCollectionIDFromEntityType(entityType));
  }
}
