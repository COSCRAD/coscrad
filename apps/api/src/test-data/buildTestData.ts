import { InMemorySnapshot, resourceTypes } from '../domain/types/resourceTypes';
import buildBookTestData from './buildBookTestData';
import buildEdgeConnectionTestData from './buildEdgeConnectionTestData';
import buildPhotographTestData from './buildPhotographTestData';
import buildSpatialFeatureTestData from './buildSpatialFeatureTestData';
import buildTagTestData from './buildTagTestData';
import buildTermTestData from './buildTermTestData';
import buildTranscribedAudioTestData from './buildTranscribedAudioTestData';
import buildVocabularyListTestData from './buildVocabularyListTestData';

/**
 * **note** When adding new test data \ modifying existing test data, be sure to
 * run `validateTestData.spec.ts` to ensure your test data satisfies all domain
 * invariants.
 */
export default (): InMemorySnapshot => ({
    resources: {
        [resourceTypes.term]: buildTermTestData(),
        [resourceTypes.vocabularyList]: buildVocabularyListTestData(),
        [resourceTypes.transcribedAudio]: buildTranscribedAudioTestData(),
        [resourceTypes.book]: buildBookTestData(),
        [resourceTypes.photograph]: buildPhotographTestData(),
        [resourceTypes.spatialFeature]: buildSpatialFeatureTestData(),
    },
    connections: buildEdgeConnectionTestData(),
    tags: buildTagTestData(),
});
