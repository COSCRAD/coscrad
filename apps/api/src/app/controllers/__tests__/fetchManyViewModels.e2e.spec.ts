import { INestApplication } from '@nestjs/common';
import * as request from 'supertest';
import { Resource } from '../../../domain/models/resource.entity';
import { InMemorySnapshotOfResources, ResourceType } from '../../../domain/types/ResourceType';
import { ArangoConnectionProvider } from '../../../persistence/database/arango-connection.provider';
import generateRandomTestDatabaseName from '../../../persistence/repositories/__tests__/generateRandomTestDatabaseName';
import TestRepositoryProvider from '../../../persistence/repositories/__tests__/TestRepositoryProvider';
import buildTestData from '../../../test-data/buildTestData';
import httpStatusCodes from '../../constants/httpStatusCodes';
import buildViewModelPathForResourceType from '../utilities/buildViewModelPathForResourceType';
import setUpIntegrationTest from './setUpIntegrationTest';

describe('When fetching multiple resources', () => {
    const testDatabaseName = generateRandomTestDatabaseName();

    let app: INestApplication;

    let arangoConnectionProvider: ArangoConnectionProvider;

    let testRepositoryProvider: TestRepositoryProvider;

    const testData = buildTestData();

    const tagTestData = testData.tags;

    const resourceTestData = testData.resources;

    const testDataWithAllResourcesPublished = Object.entries(resourceTestData).reduce(
        (accumulatedData: InMemorySnapshotOfResources, [ResourceType, instances]) => ({
            ...accumulatedData,
            [ResourceType]: instances.map((instance) =>
                instance.clone({
                    published: true,
                })
            ),
        }),
        {}
    );

    beforeAll(async () => {
        ({ app, arangoConnectionProvider, testRepositoryProvider } = await setUpIntegrationTest({
            ARANGO_DB_NAME: testDatabaseName,
            BASE_DIGITAL_ASSET_URL: 'https://www.mysound.org/downloads/',
        }));
    });

    Object.values(ResourceType).forEach((resourceType) => {
        const endpointUnderTest = `/${buildViewModelPathForResourceType(resourceType)}`;

        describe(`GET ${endpointUnderTest}`, () => {
            beforeEach(async () => {
                await testRepositoryProvider.testSetup();
            });

            afterEach(async () => {
                await testRepositoryProvider.testTeardown();
            });

            describe('when all of the resources are published', () => {
                beforeEach(async () => {
                    await testRepositoryProvider.addResourcesOfManyTypes(
                        testDataWithAllResourcesPublished
                    );

                    await testRepositoryProvider.getTagRepository().createMany(tagTestData);
                });

                it(`should fetch multiple resources of type ${resourceType}`, async () => {
                    const res = await request(app.getHttpServer()).get(
                        `/resources${endpointUnderTest}`
                    );

                    expect(res.status).toBe(httpStatusCodes.ok);

                    // TODO Support new API for all resources
                    const data = [ResourceType.song, ResourceType.term].includes(resourceType)
                        ? res.body.data
                        : res.body;

                    expect(data.length).toBe(
                        testDataWithAllResourcesPublished[resourceType].length
                    );

                    expect(res.body).toMatchSnapshot();
                });
            });

            describe(`when some of the resources are unpublished`, () => {
                /**
                 * Note that there is no requirement that the test data have
                 * `published = true`
                 */
                const publishedResourcesToAdd = resourceTestData[resourceType].map(
                    (instance: Resource) =>
                        instance.clone({
                            published: true,
                        })
                );

                const unpublishedResourcesToAdd = resourceTestData[resourceType]
                    // We want a different number of published \ unpublished terms
                    .slice(0, -1)
                    .map((instance, index) =>
                        instance.clone({
                            id: `UNPUBLISHED-00${index + 1}`,
                            published: false,
                        })
                    );

                beforeEach(async () => {
                    await testRepositoryProvider.addResourcesOfSingleType(resourceType, [
                        ...unpublishedResourcesToAdd,
                        ...publishedResourcesToAdd,
                    ]);

                    await testRepositoryProvider.getTagRepository().createMany(tagTestData);
                });

                it('should return the expected number of results', async () => {
                    const res = await request(app.getHttpServer()).get(
                        `/resources${endpointUnderTest}`
                    );

                    // TODO Support new API for all resources
                    const data = [ResourceType.song, ResourceType.term].includes(resourceType)
                        ? res.body.data
                        : res.body;

                    expect(data.length).toBe(publishedResourcesToAdd.length);

                    // Sanity check
                    expect(publishedResourcesToAdd.length).not.toEqual(
                        unpublishedResourcesToAdd.length
                    );

                    expect(res.body).toMatchSnapshot();
                });
            });
        });
    });

    afterAll(async () => {
        await arangoConnectionProvider.dropDatabaseIfExists();

        await app.close();
    });
});
