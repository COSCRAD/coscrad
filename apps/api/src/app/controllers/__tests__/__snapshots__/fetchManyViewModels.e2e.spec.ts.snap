// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When fetching multiple resources GET /bibliographicReferences when all of the resources are published should fetch multiple resources of type bibliographicReference 1`] = `
Array [
  Object {
    "data": Object {
      "abstract": "This is the abstract, NOT a general note!",
      "creators": Array [
        Object {
          "name": "Alana Duvernay",
          "type": "author",
        },
      ],
      "isbn": "978-1-895811-34-6",
      "numberOfPages": 455,
      "place": "Kennebunk Port, Maine",
      "publisher": "Atlantic Publishing",
      "title": "A Day in the Life",
      "type": "book",
      "url": "https://atlanticpublishing.com",
      "year": 1999,
    },
    "id": "1",
    "tags": Array [
      Object {
        "id": "2",
        "label": "animals",
      },
    ],
  },
  Object {
    "data": Object {
      "abstract": "An analysis of the Cariboo Chilcotin Justice inquiry.",
      "creators": Array [
        Object {
          "name": "Sigurd Purcell",
          "type": "author",
        },
      ],
      "doi": "10.14288/bcs.v0i19.784",
      "issn": "00052949",
      "issueDate": "Spring 2013",
      "pages": "109-113,115-136,240",
      "publicationTitle": "Journal of History",
      "title": "Report on the Cariboo Chilcotin Justice inquiry",
      "type": "journalArticle",
      "url": "https://search.proquest.com/docview/1682229477/abstract/7836BCEA06014582PQ/1",
    },
    "id": "23",
    "tags": Array [],
  },
]
`;

exports[`When fetching multiple resources GET /bibliographicReferences when some of the resources are unpublished should return the expected number of results 1`] = `
Array [
  Object {
    "data": Object {
      "abstract": "This is the abstract, NOT a general note!",
      "creators": Array [
        Object {
          "name": "Alana Duvernay",
          "type": "author",
        },
      ],
      "isbn": "978-1-895811-34-6",
      "numberOfPages": 455,
      "place": "Kennebunk Port, Maine",
      "publisher": "Atlantic Publishing",
      "title": "A Day in the Life",
      "type": "book",
      "url": "https://atlanticpublishing.com",
      "year": 1999,
    },
    "id": "1",
    "tags": Array [
      Object {
        "id": "2",
        "label": "animals",
      },
    ],
  },
  Object {
    "data": Object {
      "abstract": "An analysis of the Cariboo Chilcotin Justice inquiry.",
      "creators": Array [
        Object {
          "name": "Sigurd Purcell",
          "type": "author",
        },
      ],
      "doi": "10.14288/bcs.v0i19.784",
      "issn": "00052949",
      "issueDate": "Spring 2013",
      "pages": "109-113,115-136,240",
      "publicationTitle": "Journal of History",
      "title": "Report on the Cariboo Chilcotin Justice inquiry",
      "type": "journalArticle",
      "url": "https://search.proquest.com/docview/1682229477/abstract/7836BCEA06014582PQ/1",
    },
    "id": "23",
    "tags": Array [],
  },
]
`;

exports[`When fetching multiple resources GET /books when all of the resources are published should fetch multiple resources of type book 1`] = `
Array [
  Object {
    "author": "John Doeburg",
    "id": "23",
    "publicationDate": "2002",
    "subtitle": "The wolf's revenge",
    "tags": Array [
      Object {
        "id": "1",
        "label": "plants",
      },
    ],
    "title": "Three Little Pigs",
  },
  Object {
    "author": "Jane Donnaldson",
    "id": "24",
    "publicationDate": "2011",
    "tags": Array [
      Object {
        "id": "1",
        "label": "plants",
      },
    ],
    "title": "An Adventure Story",
  },
]
`;

exports[`When fetching multiple resources GET /books when some of the resources are unpublished should return the expected number of results 1`] = `
Array [
  Object {
    "author": "John Doeburg",
    "id": "23",
    "publicationDate": "2002",
    "subtitle": "The wolf's revenge",
    "tags": Array [
      Object {
        "id": "1",
        "label": "plants",
      },
    ],
    "title": "Three Little Pigs",
  },
  Object {
    "author": "Jane Donnaldson",
    "id": "24",
    "publicationDate": "2011",
    "tags": Array [
      Object {
        "id": "1",
        "label": "plants",
      },
    ],
    "title": "An Adventure Story",
  },
]
`;

exports[`When fetching multiple resources GET /photographs when all of the resources are published should fetch multiple resources of type photograph 1`] = `
Array [
  Object {
    "id": "0",
    "imageURL": "https://www.mysound.org/downloads/cat.png",
    "photographer": "Susie McRealart",
    "tags": Array [
      Object {
        "id": "3",
        "label": "placenames",
      },
    ],
  },
  Object {
    "id": "1",
    "imageURL": "https://www.mysound.org/downloads/dog.png",
    "photographer": "Robert McRealart",
    "tags": Array [
      Object {
        "id": "4",
        "label": "songs",
      },
    ],
  },
  Object {
    "id": "2",
    "imageURL": "https://www.mysound.org/downloads/wildflower.png",
    "photographer": "Kenny Tree-Huggens",
    "tags": Array [],
  },
]
`;

exports[`When fetching multiple resources GET /photographs when some of the resources are unpublished should return the expected number of results 1`] = `
Array [
  Object {
    "id": "0",
    "imageURL": "https://www.mysound.org/downloads/cat.png",
    "photographer": "Susie McRealart",
    "tags": Array [
      Object {
        "id": "3",
        "label": "placenames",
      },
    ],
  },
  Object {
    "id": "1",
    "imageURL": "https://www.mysound.org/downloads/dog.png",
    "photographer": "Robert McRealart",
    "tags": Array [
      Object {
        "id": "4",
        "label": "songs",
      },
    ],
  },
  Object {
    "id": "2",
    "imageURL": "https://www.mysound.org/downloads/wildflower.png",
    "photographer": "Kenny Tree-Huggens",
    "tags": Array [],
  },
]
`;

exports[`When fetching multiple resources GET /songs when all of the resources are published should fetch multiple resources of type song 1`] = `
Array [
  Object {
    "audioURL": "songs/lamb.mp3",
    "contributions": Array [
      "1 (performer)",
    ],
    "id": "1",
    "lengthMilliseconds": 3500,
    "lyrics": "Mary had a little lamb, little lamb.",
    "startMilliseconds": 0,
    "tags": Array [],
    "title": "Song title in language",
    "titleEnglish": "Mary had a little lamb",
  },
]
`;

exports[`When fetching multiple resources GET /songs when some of the resources are unpublished should return the expected number of results 1`] = `
Array [
  Object {
    "audioURL": "songs/lamb.mp3",
    "contributions": Array [
      "1 (performer)",
    ],
    "id": "1",
    "lengthMilliseconds": 3500,
    "lyrics": "Mary had a little lamb, little lamb.",
    "startMilliseconds": 0,
    "tags": Array [],
    "title": "Song title in language",
    "titleEnglish": "Mary had a little lamb",
  },
]
`;

exports[`When fetching multiple resources GET /spatialFeatures when all of the resources are published should fetch multiple resources of type spatialFeature 1`] = `
Array [
  Object {
    "geometry": Object {
      "coordinates": Array [
        100,
        0,
      ],
      "type": "Point",
    },
    "id": "100",
    "tags": Array [
      Object {
        "id": "1",
        "label": "plants",
      },
    ],
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        20,
        10,
      ],
      "type": "Point",
    },
    "id": "101",
    "tags": Array [
      Object {
        "id": "5",
        "label": "legends",
      },
    ],
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        70.5,
        23.5,
      ],
      "type": "Point",
    },
    "id": "102",
    "tags": Array [
      Object {
        "id": "2",
        "label": "animals",
      },
    ],
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          1,
          2,
        ],
        Array [
          1,
          3,
        ],
        Array [
          1.5,
          3.5,
        ],
        Array [
          3,
          4.2,
        ],
      ],
      "type": "LineString",
    },
    "id": "1",
    "tags": Array [],
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            1,
            3,
          ],
          Array [
            2,
            5,
          ],
          Array [
            1.5,
            3.7,
          ],
          Array [
            1.2,
            3.3,
          ],
          Array [
            1,
            3,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "id": "300",
    "tags": Array [],
  },
]
`;

exports[`When fetching multiple resources GET /spatialFeatures when some of the resources are unpublished should return the expected number of results 1`] = `
Array [
  Object {
    "geometry": Object {
      "coordinates": Array [
        100,
        0,
      ],
      "type": "Point",
    },
    "id": "100",
    "tags": Array [
      Object {
        "id": "1",
        "label": "plants",
      },
    ],
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        20,
        10,
      ],
      "type": "Point",
    },
    "id": "101",
    "tags": Array [
      Object {
        "id": "5",
        "label": "legends",
      },
    ],
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        70.5,
        23.5,
      ],
      "type": "Point",
    },
    "id": "102",
    "tags": Array [
      Object {
        "id": "2",
        "label": "animals",
      },
    ],
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          1,
          2,
        ],
        Array [
          1,
          3,
        ],
        Array [
          1.5,
          3.5,
        ],
        Array [
          3,
          4.2,
        ],
      ],
      "type": "LineString",
    },
    "id": "1",
    "tags": Array [],
  },
  Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            1,
            3,
          ],
          Array [
            2,
            5,
          ],
          Array [
            1.5,
            3.7,
          ],
          Array [
            1.2,
            3.3,
          ],
          Array [
            1,
            3,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "id": "300",
    "tags": Array [],
  },
]
`;

exports[`When fetching multiple resources GET /terms when all of the resources are published should fetch multiple resources of type term 1`] = `
Array [
  Object {
    "contributor": "",
    "id": "1",
    "tags": Array [
      Object {
        "id": "1",
        "label": "plants",
      },
    ],
    "term": "Chil-term-1",
    "termEnglish": "Engl-term-1",
  },
  Object {
    "contributor": "",
    "id": "2",
    "tags": Array [
      Object {
        "id": "2",
        "label": "animals",
      },
    ],
    "term": "Chil-term-2",
    "termEnglish": "Engl-term-2",
  },
  Object {
    "contributor": "",
    "id": "3",
    "tags": Array [
      Object {
        "id": "2",
        "label": "animals",
      },
    ],
    "term": "Chil-term-no-english",
  },
]
`;

exports[`When fetching multiple resources GET /terms when some of the resources are unpublished should return the expected number of results 1`] = `
Array [
  Object {
    "contributor": "",
    "id": "1",
    "tags": Array [
      Object {
        "id": "1",
        "label": "plants",
      },
    ],
    "term": "Chil-term-1",
    "termEnglish": "Engl-term-1",
  },
  Object {
    "contributor": "",
    "id": "2",
    "tags": Array [
      Object {
        "id": "2",
        "label": "animals",
      },
    ],
    "term": "Chil-term-2",
    "termEnglish": "Engl-term-2",
  },
  Object {
    "contributor": "",
    "id": "3",
    "tags": Array [
      Object {
        "id": "2",
        "label": "animals",
      },
    ],
    "term": "Chil-term-no-english",
  },
]
`;

exports[`When fetching multiple resources GET /transcribedAudioItems when all of the resources are published should fetch multiple resources of type transcribedAudio 1`] = `
Array [
  Object {
    "audioURL": "https://www.mysound.org/downloads/123.mp3.mp3",
    "id": "110",
    "length": 20000,
    "plainText": "
[12000] There once was a little wooden boy. [15550]
[18300] His name was Pinocchio. [19240]",
    "start": 0,
    "tags": Array [
      Object {
        "id": "3",
        "label": "placenames",
      },
    ],
  },
]
`;

exports[`When fetching multiple resources GET /transcribedAudioItems when some of the resources are unpublished should return the expected number of results 1`] = `
Array [
  Object {
    "audioURL": "https://www.mysound.org/downloads/123.mp3.mp3",
    "id": "110",
    "length": 20000,
    "plainText": "
[12000] There once was a little wooden boy. [15550]
[18300] His name was Pinocchio. [19240]",
    "start": 0,
    "tags": Array [
      Object {
        "id": "3",
        "label": "placenames",
      },
    ],
  },
]
`;

exports[`When fetching multiple resources GET /vocabularyLists when all of the resources are published should fetch multiple resources of type vocabularyList 1`] = `
Array [
  Object {
    "entries": Array [
      Object {
        "term": Object {
          "contributor": "",
          "id": "1",
          "term": "Chil-term-1",
          "termEnglish": "Engl-term-1",
        },
        "variableValues": Object {
          "person": "11",
        },
      },
      Object {
        "term": Object {
          "contributor": "",
          "id": "2",
          "term": "Chil-term-2",
          "termEnglish": "Engl-term-2",
        },
        "variableValues": Object {
          "person": "21",
        },
      },
    ],
    "id": "1",
    "name": "test VL 1 chil",
    "nameEnglish": "test VL 1 engl",
    "tags": Array [
      Object {
        "id": "1",
        "label": "plants",
      },
    ],
    "variables": Array [
      Object {
        "name": "person",
        "type": "dropbox",
        "validValues": Array [
          Object {
            "display": "I",
            "value": "11",
          },
          Object {
            "display": "We",
            "value": "21",
          },
        ],
      },
    ],
  },
  Object {
    "entries": Array [
      Object {
        "term": Object {
          "contributor": "",
          "id": "2",
          "term": "Chil-term-2",
          "termEnglish": "Engl-term-2",
        },
        "variableValues": Object {
          "person": "23",
        },
      },
    ],
    "id": "2",
    "name": "test VL 2 CHIL- no engl name",
    "tags": Array [
      Object {
        "id": "5",
        "label": "legends",
      },
    ],
    "variables": Array [],
  },
]
`;

exports[`When fetching multiple resources GET /vocabularyLists when some of the resources are unpublished should return the expected number of results 1`] = `
Array [
  Object {
    "entries": Array [],
    "id": "1",
    "name": "test VL 1 chil",
    "nameEnglish": "test VL 1 engl",
    "tags": Array [
      Object {
        "id": "1",
        "label": "plants",
      },
    ],
    "variables": Array [
      Object {
        "name": "person",
        "type": "dropbox",
        "validValues": Array [
          Object {
            "display": "I",
            "value": "11",
          },
          Object {
            "display": "We",
            "value": "21",
          },
        ],
      },
    ],
  },
  Object {
    "entries": Array [],
    "id": "2",
    "name": "test VL 2 CHIL- no engl name",
    "tags": Array [
      Object {
        "id": "5",
        "label": "legends",
      },
    ],
    "variables": Array [],
  },
]
`;
